### PR TITLE
CUMULUS-1337: s3 replicator lambda tf module

### PR DESCRIPTION
**Pull request by nemreid**
**Summary:** Summary of changes

Supports [CUMULUS-1337](https://bugs.earthdata.nasa.gov/browse/CUMULUS-1337)

## Changes

* Adds lambda function to replicate created objects to a secondary bucket.
* Adds Terraform configuration to deploy lambda function as a stand-alone module.

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Adhoc testing
- [ ] Integration tests

### CODE REVIEW

File: packages/s3-replicator/test/test-index.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const test = require('ava');
+
+const { handler } = require('../index');
+
+test('handler returns immediately on non-create event', async (t) => {
```
Comment (nemreid) [inline packages/s3-replicator/test/test-index.js]: Wouldn’t mind some feedback on how to write a good unit test for the do-copy case, considering it’s really just a single `copyObject` call.

File: packages/s3-replicator/iam.tf [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+resource "aws_iam_role" "replicator_lambda_role" {
+  name                 = "${var.prefix}_replicator_lambda_role"
+  assume_role_policy   = data.aws_iam_policy_document.assume_lambda_role.json
+  permissions_boundary = var.permissions_boundary
+}
+
+data "aws_iam_policy_document" "replicator_policy_document" {
+  statement {
```
Comment (flamingbear) [inline packages/s3-replicator/iam.tf]: Do we need all of these permissions?  I would guess ec2 and logs are not needed?

File: packages/s3-replicator/README.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+aws_profile          = "ngap"                 # AWS profile used for deployment
+permissions_boundary = "NGAPShRoleBoundary"   # IAM permissions boundary
+vpc_id               = "vpc-123456"           # VPC ID (for Lambda)
+subnet_ids           = ["subnet-123456"]      # Subnet IDs (for Lambda)
+security_groups      = ["sg-123456"]          # Security Group IDs (for Lambda)
+source_bucket        = "source-bucket-name"   # Source S3 bucket
+source_prefix        = "source-prefix"        # Source object prefix e.g. 'path/to/filedir'
+target_bucket        = "target-bucket-name"   # Target S3 bucket e.g. 'mybucket'
+target_prefix        = "target-prefix"        # Target object prefix path
+```
```
Comment (flamingbear) [inline packages/s3-replicator/README.md]: Probably need some basic instructions on how to deploy, or point to basic instructions.

File: packages/s3-replicator/cfg.tfvars.sample [hunk 4]
```diff
@@ -0,0 +1,10 @@
+prefix               = "myprefix"
+aws_profile          = "ngap"
+permissions_boundary = "NGAPShRoleBoundary"
```
Comment (flamingbear) [inline packages/s3-replicator/cfg.tfvars.sample]: should look like an ARN.

File: packages/s3-replicator/README.md (hunk 3 above)
Comment (flamingbear) [reply packages/s3-replicator/README.md]: I'm unable to deploy and I'm sure I've got something configured wrong.  getting errors about my user not having permissions

```Error: Error creating IAM Role mhs3_replicator_lambda_role: AccessDenied: User: arn:aws:iam::123456789012:user/CumulusSandboxAppDeveloper-mhsavoie-137 is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::123456789012:role/mhs3_replicator_lambda_role ```

Comment (nemreid) [reply packages/s3-replicator/README.md]: Do you have AWS credentials in your environment variables? They will override your `aws_profile`  in the `.tfvars` file, so make sure they aren’t for a different account or something. I’ll add a note about what command to deploy with.

File: packages/s3-replicator/cfg.tfvars.sample (hunk 4 above)
Comment (nemreid) [reply packages/s3-replicator/cfg.tfvars.sample]: Good catch 👍

File: packages/s3-replicator/iam.tf (hunk 2 above)
Comment (nemreid) [reply packages/s3-replicator/iam.tf]: EC2 network interface permissions are necessary to run the lambda in a VPC. Logs perms are needed to write cloudwatch logs for error logging. The latter is optional I guess, but in that case there will be no logging whatsoever.

File: packages/s3-replicator/README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+NOTE: Terraform will ignore the `aws_profile` config variable if you have static credentials or environment variables set, see the [AWS Provider page](https://www.terraform.io/docs/providers/aws/index.html#authentication).
+
+## Configuration
+
+Configuration variables are shown in `cfg.tfvars.sample`, and are explained below:
+
+```
+prefix               = "myprefix"             # prefix to name created replicator resources
+aws_profile          = "ngap"                 # AWS profile used for deployment
+permissions_boundary = "NGAPShRoleBoundary"   # IAM permissions boundary
```
Comment (flamingbear) [inline packages/s3-replicator/README.md]:
```suggestion
permissions_boundary = "NGAPShRoleBoundaryArn" # IAM permissions boundary
```

File: packages/s3-replicator/test/test-index.js (hunk 1 above)
Comment (flamingbear) [reply packages/s3-replicator/test/test-index.js]: I don't have a good test except to test the implementation, which isn't to useful.

**Comment by flamingbear** [comment]
I think this is a good start, I'll be sure to include it in the docs on setting up logging for an NGAP deployment.

**Pull request merged**